site/docs/how-tos/grpc.md

```diff
@@ -0,0 +1,183 @@
+---
+title: Call gRPC service
+description: How to call unary or bidirectional streaming gRPC service from a pipeline.
+---
+
+The [Stateless Transforms](../api/stateless-transforms.md#mapusingservice)
+section shows how to call and use services to transform items in the
+pipeline. One way to implement a remote service is using
+[gRPC](https://grpc.io/) - an open source universal RPC framework, which
+is available for many platforms and languages.
+
+The `hazelcast-jet-grpc` module makes it easy to perform calls to a
+gRPC service. Currently, two types of gRPC services are supported:
+
+- unary service
+- bidirectional streaming service.
+
+This how to guide shows how to use these services in your pipeline.
+
+## Module dependency
+
+Add a dependency on `hazelcast-jet-grpc` module to use the gRPC services
+
+```xml
+<dependency>
+    <groupId>com.hazelcast.jet</groupId>
+    <artifactId>hazelcast-jet-grpc</artifactId>
+    <version>4.1</version>
+</dependency>
+```
+
+## Unary service
+
+A unary service allows to send a single request and receive a response,
+similar to a normal function call.
+
+For example, given the protobuf definition below
+
+```proto
+service ProductService {
+  rpc ProductInfo (ProductInfoRequest) returns (ProductInfoReply) {}
+}
+```
+
+We can create the following service factory using
+`GrpcServices.unaryService()` method:
+
+```java
+ServiceFactory<?, ? extends GrpcService<ProductInfoRequest, ProductInfoReply>> productService = unaryService(
+    () -> ManagedChannelBuilder.forAddress("localhost", PORT).useTransportSecurity().usePlaintext(),
+    channel -> ProductServiceGrpc.newStub(channel)::productInfo
+);
+```
+
+The first parameter is a supplier returning a channel builder. Modify
+the builder settings as required, e.g. to enable TLS via
+`io.grpc.ManagedChannelBuilder.useTransportSecurity()`.  
+
+The second parameter is a function which, given a channel, creates the
+stub (the `ProductServiceGrpc` class is auto-generated by the protobuf
+compiler) and returns a function that calls the stub given
+the input item and the observer. Again you may modify the stub according
+to your needs. The returned function will be called once per input item.
+
+The full type of the second parameter is as follows (with wildcards
+omitted):
+
+```java
+FunctionEx<ManagedChannel, BiConsumerEx<T, StreamObserver<R>>> callStubFn
+```
+
+Now the service factory can be used in any of the `mapUsingService*`
+methods, preferably the `mapUsingServiceAsync`.
+
+```java
+StreamStage<Trade> trades = ...
+trades.mapUsingServiceAsync(productService,
+(service, trade) -> {
+    ProductInfoRequest request = ProductInfoRequest.newBuilder().setId(trade.productId()).build();
+    return service.call(request).thenApply(productReply -> tuple2(trade, productReply.getProductName()));
+})
+```
+
+## Bidirectional streaming service
```

Comment:
for bidirectional we may show the server-side implementation as well, because it's not immediately obvious perhaps? just as a snippet, we probably don't need the full code.

Reply:
I think the user either already knows gRPC and just wants to use it from Jet, or doesn't and in that case would need to follow some other tutorial to create the server. A snippet from e.g. here wouldn't be enough

https://github.com/hazelcast/hazelcast-jet/blob/dbbe962fc14043f39df80eb72371643b7adde5d6/examples/grpc/src/main/java/com/hazelcast/jet/examples/grpc/BrokerServiceImpl.java#L38